base/sparse/sparsematrix.jl

```diff
@@ -3380,40 +3380,38 @@ function expandptr(V::Vector{<:Integer})
     res
 end
 
-## diag and related using an iterator
 
-mutable struct SpDiagIterator{Tv,Ti}
-    A::SparseMatrixCSC{Tv,Ti}
-    n::Int
-end
-SpDiagIterator(A::SparseMatrixCSC) = SpDiagIterator(A,minimum(size(A)))
-
-length(d::SpDiagIterator) = d.n
-start(d::SpDiagIterator) = 1
-done(d::SpDiagIterator, j) = j > d.n
-
-function next(d::SpDiagIterator{Tv}, j) where Tv
-    A = d.A
-    r1 = Int(A.colptr[j])
-    r2 = Int(A.colptr[j+1]-1)
-    (r1 > r2) && (return (zero(Tv), j+1))
-    r1 = searchsortedfirst(A.rowval, j, r1, r2, Forward)
-    (((r1 > r2) || (A.rowval[r1] != j)) ? zero(Tv) : A.nzval[r1], j+1)
+function diag(A::SparseMatrixCSC{Tv,Ti}, d::Integer=0) where {Tv,Ti}
+    m, n = size(A)
+    if !(-m <= d <= n)
+        throw(ArgumentError("requested diagonal, $d, out of bounds in matrix of size ($m, $n)"))
+    end
+    l = d < 0 ? min(m+d,n) : min(n-d,m)
+    r, c = d <= 0 ? (-d, 0) : (0, d) # start row/col -1
```

Comment:
Perhaps either type `r` and `c` or use `zero` to on the right side to avoid potential type instability?